src/Http/Http.Extensions/gen/RequestDelegateGenerator.cs

```diff
@@ -0,0 +1,149 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Linq;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.AspNetCore.Http.SourceGeneration.StaticRouteHandlerModel;
+
+namespace Microsoft.AspNetCore.Http.SourceGeneration;
+
+[Generator]
+public class RequestDelegateGenerator : IIncrementalGenerator
+{
+    private static readonly string[] _knownMethods =
+    {
+        "MapGet",
+        "MapPost",
+        "MapPut",
+        "MapDelete",
+        "MapPatch",
+        "Map",
+    };
+
+    public void Initialize(IncrementalGeneratorInitializationContext context)
+    {
+        var mapActionOperations = context.SyntaxProvider.CreateSyntaxProvider(
+            predicate: (node, _) => node is InvocationExpressionSyntax
+            {
+                Expression: MemberAccessExpressionSyntax
+                {
+                    Name: IdentifierNameSyntax
+                    {
+                        Identifier: { ValueText: var method }
+                    }
+                },
+                ArgumentList: { Arguments: { Count: 2 } args }
+            } && _knownMethods.Contains(method),
+            transform: (context, token) => context.SemanticModel.GetOperation(context.Node, token) as IInvocationOperation);
```

Comment:
I think this is not good for incrementality. I think it's better to call `StaticRouteHandlerModelParser.GetEndpointFromOperation` here in the transform instead of a different step. But will let @333fred or @CyrusNajmabadi confirm that.

Note that currently it seems that all the models are regular classes that aren't equatable, which might break caching. Best to use `record` (or implement a proper equality)

Reply:
> I think this is not good for incrementality. I think it's better to call StaticRouteHandlerModelParser.GetEndpointFromOperation here in the transform instead of a different step. But will let @333fred or @CyrusNajmabadi confirm that.

That's fair. Is the idea that if we compute the model in the transform then we avoid running the pipeline to convert operations to `EndpointModel`s each time there is a change in syntax even if the operation hasn't changed since operations are immutable?

> Note that currently it seems that all the models are regular classes that aren't equatable, which might break caching. Best to use record (or implement a proper equality)

Yep, eventually the `StaticHandlerModel` will implement equality checks on all the types. The idea is to evolve the model as we build out features so there's not much to do in the realm of checks at the moment (we only support one type of delegate signature).

Reply:
@captainsafia There are several concerns with having IOperation or symbols as part of the pipeline. Some are about memory usage, and some about proper incrementality (ie, avoid running the whole pipeline every time the user types a character). See https://github.com/dotnet/roslyn-analyzers/issues/6352

If this PR will ship to end-users, I think it's better to both fix the pipeline and convert all the data models to records.

Reply:
> If this PR will ship to end-users, I think it's better to both fix the pipeline and convert all the data models to records.

It will eventually ship to end-users but it'll be off by default so we've got wiggle room to refine as needed.

> I think it's better to both fix the pipeline and convert all the data models to records.

I assume it is sufficient to implement `IEquatable` on the static model as an alternative to using records.

Can you clarify what you mean by "fix the pipeline"? I read through the thread and it seems to indicate that we shouldn't be using symbols or operations anywhere at all in the generator which makes the implementation difficult, if not impossible. Did I understand the guidance correctly?

Reply:
Yes I think implementing `IEquatable` should be sufficient, though records are just easier.

> it seems to indicate that we shouldn't be using symbols or operations anywhere at all in the generator

You can only use it as the first step, i.e, the `transform` of `CreateSyntaxProvider` should return the data model directly.

I see you already done that in b9b8cec4d0ba3f6df1b77a620805e849682ee387, the remaining part here is to implement equality for the data models.

Otherwise, they will always compare unequal and the generator will run completely with every typed character in IDE.

Reply:
https://github.com/dotnet/aspnetcore/pull/45924/commits/bb9243d3b097236f6a094ccc7e24ab05fcc60bc0 implements the `IEquatable` interface on the model. These implementations should evolve as we build out the bottle but this should be a sufficient start.